modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -2487,6 +2488,79 @@ def _callable_func(self, func, axis, *args, **kwargs):
     groupby_prod = GroupByReduce.register("prod")
     groupby_sum = GroupByReduce.register("sum")
 
+    @staticmethod
+    def _groupby_mean_reduce(dfgb, **kwargs):
+        """
+        Compute mean value in each group using sums/counts values from map stage.
+
+        Parameters
+        ----------
+        dfgb : pandas.DataFrameGroupBy
+            GroupBy object for column-partition.
+        **kwargs : dict
+            Additional keyword parameters to be passed in ``pandas.DataFrameGroupBy.sum``.
+
+        Returns
+        -------
+        pandas.DataFrame
+            A pandas Dataframe with mean values in each column of each group.
+        """
+        sums_counts_df = dfgb.sum(min_count=1, **kwargs)
+        sum_df = sums_counts_df.iloc[:, : len(sums_counts_df.columns) // 2]
+        count_df = sums_counts_df.iloc[:, len(sums_counts_df.columns) // 2 :]
+        return sum_df / count_df
+
+    def groupby_mean(self, by, axis, groupby_kwargs, agg_args, agg_kwargs, drop=False):
+        numeric_only = agg_kwargs.get("numeric_only", False)
+        datetime_cols = (
+            {
+                col: dtype
+                for col, dtype in dict(self.dtypes).items()
+                if is_datetime64_any_dtype(dtype)
+            }
+            if not numeric_only
+            else dict()
+        )
+
+        if len(datetime_cols) > 0:
+            datetime_qc = self.getitem_array(datetime_cols)
+            if datetime_qc.isna().any().any(axis=1).to_pandas().squeeze():
```

Comment:
I've checked the time of this condition. As expected, it significantly depends from the count of datetime-like columns in original dataframe. If number of rows will be huge the new implementation will show better times than current (I'm limited in hardware to check this on more number of rows).


Performance result of `groupby(["Col1"], as_index=True).mean(numeric_only=False)`, MODIN_NPARTITIONS=4:

| shape (rows x cols), s| 10m x 16 (15 datetime columns) | 10m x 70(69 datetime columns) | 10m x 16 (2 datetime columns) | 10m x 70 (2datetime columns)| 
| ----------- | :-----------: | :-----------: | :-----------: | :-----------: |
|condition time |0.71|1.4|0.07|0.07|
|modin #4591 |2.3|5.79|0.89|2.9|
|modin master |3.21|4.08|1.11|2.41|

I don't think that dataset which contains only datetime columns the most used case. But I agree, in this case we will have the bottleneck in case small number of rows. If you have any suggestions how we can check `datetimce_qc`  on containing of at least one `NaT` value  please share :-)

Reply:
Given the note you linked in the description:

> EXCEPTION case: If data contains at least one datetime-like column with NaT values the execution will use previous implementation (broadcast_apply_full_axis). The reason: it's unclear how to exclude NaT values from processing of groupby.sum on map-stage, because all datetime-like data must be converted to numeric type to make possible using of groupby.sum. Possible, this case should be covered in a separate issue.

What if we perform `result.dropna` after `result.astype`? We should get the result without nans in that case.

Reply:
What if we fill nans with 0 to make it possible to use groupby.sum?

Reply:
In this case we will corrupt `mean` result. Because `mean([2, nan]) = 2`, but `mean([2,0]=1)`

Reply:
Could you elaborate why `broadcast_apply_full_axis` handles the case, but `tree_reduce` not?

Reply:
Old approach (`broadcast_apply_full_axis`) applyes `pandas.groupby.mean` to entire column partitions. `pandas.groupby.mean` correctly processes datetimelike columns. The new approach uses `groupby.sum` which can't work with datetime columns that's why we make `astype` to convert the datetime columns to integer and we have to check the NaNs.

Reply:
Is it possible to replace `pd.NaT` values to `pd.NaN` and other datetime values convert to floats (datetime->int->float) when performing map phase? groupby.sum shouldn't consider nans then.

Reply:
Ok, we can leave such an experiment in a separate issue. However, let's try to reduce amount of remote calls by using `datetime_qc.dropna().index > 0` to see if is beneficial.

Reply:
Using of `dropna` will produce wrong result because entire row/column with NaN will be dropped. In this case value of `groupby.mean` for other columns will be incorrect.

Reply:
I propose to use `dropna` just in the condition to check if there are nans.

Reply:
The next condition has checked:
```suggestion
            # Indicate, that some of rows have been dropped
            if len(datetime_qc.dropna().index) < len(self.index):
```

Performance result of `groupby(["Col1"], as_index=True).mean(numeric_only=False)`, MODIN_NPARTITIONS=4:

| shape (rows x cols), s| 10m x 16 (15 datetime columns) | 10m x 70(69 datetime columns) | 10m x 16 (2 datetime columns) | 10m x 70 (2datetime columns)| 
| ----------- | :-----------: | :-----------: | :-----------: | :-----------: |
|condition (with dropna) time |2.06|14.16|0.25|0.29|
|modin #4591(with dropna condition) |3.58|19.84|1.37|3.57|
|condition time |0.71|1.4|0.07|0.07|
|modin #4591 |2.3|5.79|0.89|2.9|
|modin master |3.21|4.08|1.11|2.41|

`dropna` signifanctly slows execution (possible, because it's axis-wise operation)

Reply:
Ok. Can we just add `not numeric_only` to `len(datetime_cols) > 0` condition above and call `super().groupby_mean` right away because it seems the flow outside of this condition performs groupby for `numeric_only=True`?

Reply:
I see. Then we don't perform GroupByReduce for datetime in case `numeric_only=False`.

Reply:
Ok, let's leave as is. Meanwhile, could you ask a question in pandas GH regarding how is it possible to handle the case of datetime values containing nans?

Also, would you be able to experiment with that I proposed in a separate issue?

> Is it possible to replace pd.NaT values to pd.NaN and convert other datetime values to floats (datetime->int->float) when performing map phase? groupby.sum shouldn't consider nans then.
> 
> We can convert only entire column at time. I don't think that enumerating over column and converting of each value to float64/pd.NA is effective. Besides, on map phase we already work with groupby object.

I think we can iterate over groups and perform necessary stuff.

Reply:
> Ok, let's leave as is. Meanwhile, could you ask a question in pandas GH regarding how is it possible to handle the case of datetime values containing nans?

A better question would be why converting of NaN/NaT value from datetime column to int produces limit value for `int` (example here https://github.com/modin-project/modin/pull/3586#discussion_r791197418) but not `None`. We anyway should convert datetime data to numeric because `pandas.groupby.sum` doesn't support datetime columns and drops them. I will ask pandas.

> Also, would you be able to experiment with that I proposed in a separate issue?

I created the issue #4626 to experiment with this case in the future.

> I think we can iterate over groups and perform necessary stuff.

Yes, we can iterate over groups in groupby object. But I still think "handmade" converting of datetime columns of group to int/float will be too expensive (so, #4626 will be place for the discussion).

Reply:
cc @prutskov, please explore this.